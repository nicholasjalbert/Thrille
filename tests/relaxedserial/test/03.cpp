//try to schedule thread 1 too many times, should auto switch to thread 0
//when thread 1 becomes disabled
#include <errno.h>
#include <pthread.h>
#include <iostream>
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>

using namespace std;

int x;
pthread_t child;
pthread_mutex_t mutex;

void builtintest() {
    printf("My return addr %p\n", __builtin_return_address(0));
}

void setup() {
    pthread_mutex_init(&mutex, NULL);
}

void teardown() {
    pthread_mutex_destroy(&mutex);
}

void * routine1(void * arg) {
    pthread_mutex_lock(&mutex);
    x = 1;
    pthread_mutex_unlock(&mutex);
}

int main(int argc, char *argv[]) {
    setup();
    builtintest();
    pthread_create(&child, NULL, routine1, NULL);
    pthread_mutex_lock(&mutex);
    x = 3;
    pthread_mutex_unlock(&mutex);
    printf("x is %d\n", x);
    pthread_join(child, NULL);
    teardown();
}

