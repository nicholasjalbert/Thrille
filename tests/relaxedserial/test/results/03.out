Thrille:Starting Default thriller...
Thrille:Starting Serializer Thriller...
Thrille:Starting *relaxed* serializer handler...
Logger: Relaxed Logger engaged
My return addr 0x400e3f
Thread 0 has run out of times to be scheduled
Thread 0 was scheduled 47 times
NonPreemptive_Context_Switch(from, to, id_addr):0:1:0x400ec6
NonPreemptive_Context_Switch(from, to, id_addr):1:0:0x400d30
Number of Context Switches: 2
Number of Preemptions: 0
